keps/sig-node/2033-rootless/README.md

```diff
@@ -0,0 +1,818 @@
+```console
+$ rootlesskit --net=slirp4netns --copy-up=/etc --copy-up=/run --copy-up=/var --pidns --cgroupns --ipcns --utsns -- containerd &
+$ nsenter -t $ROOTLESSKIT_CHILD_PID -a kubelet ... &
+$ nsenter -t $ROOTLESSKIT_CHILD_PID -a kube-proxy ... &
+$ nsenter -t $ROOTLESSKIT_CHILD_PID -a flanneld ... &
+```
+
+#### Paths
+
+Some paths like `/var/log/pods` are hardcoded in Kubernetes and hard to change.
+
+Although these directories are not writable by unprivileged users, Kubernetes does NOT need to be changed to use unprivileged home directories,
+because RootlessKit can bind-mount writable directories on these paths without the root privileges. (`rootlesskit --copy-up=/var`)
+
+#### Network
+The node components need to be executed in RootlessKit's network namespace, because an unprivileged user cannot do privileged operations in the host network namespace.
+As the components are executed inside a network namespace, `NodePorts` are not directly accessible from other hosts.
+
+An external controller should watch changes on `corev1.Service` resources and call [RootlessKit API](https://github.com/rootless-containers/rootlesskit/blob/v0.11.1/pkg/api/openapi.yaml) to set up port forwarding for the node ports.
+
+k3s implementation: https://github.com/rancher/k3s/blob/v1.17.2+k3s1/pkg/rootlessports/controller.go#L92-L96
+
+#### RootlessKit network drivers
+
+RootlessKit supports two kinds of network stacks:
+* TAP with pure usermode network stack (either `slirp4netns` or VPNKit)
+* vEth with setuid binary `lxc-user-nic`
+
+`slirp4netns` is preferred for security, `lxc-user-nic` is preferred for performance.
+
+These stacks are used for the namespace where the node components are executed in, not for the containers' namespaces.
+CNI plugins such as Flannel are expected to be used for the containers' namespace.
+
+#### CNI plugins
+
+Flannel (VXLAN) is known to work.
+
+#### cgroup
+
+[cgroup v2](../20191118-cgroups-v2.md) and systemd are required. cgroup v1 won't be supported due to security concerns.
+
+containerd supports cgroup v2 for rootless mode since containerd v1.4.
+[The master branch of CRI-O](https://github.com/cri-o/cri-o/commit/d3dbaec060e33870e5cb5c3f7ec4207837804b00) also supports cgroup v2 for rootless mode.
+It will be included in CRI-O v1.22.
+
+No code change is required on kubelet for managing cgroups, because we can use cgroup namespaces along with mount namespaces for creating writable `/sys/fs/cgroup` filesystem.
+
+### Required changes to Kubernetes
+
+Most things do work without modifying Kubernetes. But we need to modify a just few lines of kubelet and kube-proxy to ignore errors during setting some sysctl and rlimit values.
+
+#### kubelet
+Patch: ["kubelet/cm: ignore sysctl error when running in userns"](https://github.com/rootless-containers/usernetes/blob/v20210303.0/src/patches/kubernetes/0001-kubelet-cm-ignore-sysctl-error-when-running-in-usern.patch)
```

Comment:
i view this patch as similar to `--fail-swap-on` and so it seems fine to support further iteration in this space.

Reply:
Yes, thanks